Add Tld Validation Schemas to Data Layer

deBhal opened the pull request:
This PR adds mechanisms to load JSON Schemas to validation TLD contact info to facilitate #21524. We're just dumping the schemas into the state in this PR, there will be followup changes to the forms themselves to unify around using them.

To test:
1) apply D9157-code to the backend (or maybe don't first to check the error notification)
2) Domains -> Add -> Add a domain with custom validation logic - 'fr', 'uk' or 'ca'
3) Using the redux devtools or `getState().domains.management.validationSchemas.uk` Verify that the schema is loaded into the redux store.

ramonjd commented on client/state/data-layer/wpcom/domains/validation-schemas/index.js:
```diff
@@ -0,0 +1,69 @@
+/** @format */
```
[From the docs](https://github.com/Automattic/wp-calypso/tree/master/client/state/data-layer#file-structure)  I understand that file structure in the data-layer should reflect the wpcom api path. Something like `/domains/validation-schema/${ tld }` should be represented as `client/state/data-layer/wpcom/domains/validation-schema/index.js`.

ramonjd commented on client/components/data/query-tld-validation-schema/index.jsx:
```diff
@@ -0,0 +1,52 @@
+	};
+
+	componentWillMount() {
+		this.fetchMissingSchemas( this.props.tlds );
```
I believe the general advice is to make calls to remote endpoints in `componentDidMount`. `componentWillMount` is also executed on server side rendering, therefore 2 x. 

https://reactjs.org/docs/react-component.html#mounting-componentdidmount

dmsnell commented on client/components/data/query-tld-validation-schema/index.jsx:
```diff
@@ -0,0 +1,52 @@
+		this.fetchMissingSchemas( this.props.tlds );
+	}
+
+	componentWillReceiveProps( nextProps ) {
```
fun fact!

if you destructure in the arguments you won't need a separate declaration for `newTlds`

```js
componentWillReceiveProps( { tlds } ) {
  if ( ! isEqual( this.props.tlds, tlds ) ) {
    this.fetchMissingSchemas( tlds );
  }
}
```

also as a note we probably would prefer to do this in `componentDidUpdate()` in order to keep side-effects out of `willMount` / `willReceiveProps` - I'm fuzzy on this but I believe @gwwar recently informed me that there can be troubles otherwise.

gwwar replied:
Let's see current recs are to avoid `setState` in *Did to avoid unnecessary renders but it can be a good place for DOM manipulation/side-effects.
https://reactjs.org/docs/react-component.html#componentwillmount
https://gist.github.com/bvaughn/923dffb2cd9504ee440791fade8db5f9

I think dispatching redux actions eventually triggers a setState call, but I think we're still better off treating redux dispatches as a sid- effect since not all of our actions are idempotent. (eg componentWillMount in some cases might be called multiple times).

dmsnell replied:
Also I think that we’re looking at setState calls that are coming from the outside and probably after some appreciable time. That is, if it’s a Redux dispatch it will cause a re-render regardless if it’s in will or did

deBhal replied:
> Also I think that we’re looking at setState calls that are coming from the outside and probably after some appreciable time. 

Yeah, `fetchMissingSchemas()` won't trigger any props changes until the request comes back, and no `setState()`s at all, so `componentDidUpdate()` seems the right place for it.

Thanks for the cheatsheet link! - I haven't seen that before, and I love it :D

dmsnell commented on client/state/data-layer/wpcom/domains/validation-schema/index.js:
```diff
@@ -0,0 +1,69 @@
+				action
+			)
+		)
+	);
```
Is there a reason we can't batch these into a single network request? It seems like it would end up faster and use less battery power if we could do `/domains/validation-schema/?tlds=${ tlds.join( ',' ) }` or something like that

dmsnell commented on client/state/action-types.js:
```diff
@@ -143,6 +143,11 @@ export const DOMAIN_MANAGEMENT_CONTACT_DETAILS_CACHE_REQUEST_SUCCESS =
+export const DOMAIN_MANAGEMENT_VALIDATION_SCHEMA_ADD = 'DOMAIN_MANAGEMENT_VALIDATION_SCHEMA_ADD';
+export const DOMAIN_MANAGEMENT_VALIDATION_SCHEMA_REQUEST =
+	'DOMAIN_MANAGEMENT_VALIDATION_SCHEMA_REQUEST';
+export const DOMAIN_MANAGEMENT_VALIDATION_SCHEMA_REQUEST_FAILURE =
```
It doesn't look like we ever reference this. We can probably take it out.

dmsnell commented on client/state/data-layer/wpcom/domains/validation-schema/index.js:
```diff
@@ -0,0 +1,69 @@
+
+export default {
+	[ DOMAIN_MANAGEMENT_VALIDATION_SCHEMA_REQUEST ]: [
+		dispatchRequest( fetchValidationSchema, addValidationSchemas, validationSchemaFailure ),
```
also check out `dispatchRequestEx` which is very useful to simplify the code here some. instead of dispatching the actions yourself these functions can just return one action or an array of actions and each of those will be dispatched.

```js
const fetchValidationSchema = ( { tlds } ) => tlds.map( old => http(
  {
    apiVersion: '1.1',
    method: 'GET',
    path: `/domains/…`
  },
  action
) );

dispatchRequestEx( {
  fetch: fetchValidationSchema,
  onSucces: …
} )
```

dmsnell commented on client/state/data-layer/wpcom/domains/validation-schema/index.js:
```diff
@@ -0,0 +1,69 @@
+	forEach( schemas, ( schema, tld ) => {
+		dispatch( addValidationSchema( { [ tld ]: schema } ) );
+	} );
+};
```
same comment here - is there a reason we shouldn't `addValidationSchemas( tlds )` here so we can update state in one single dispatch? I don't know how many schemas we'll be dealing with but for each one as the code is written, every reducer runs and memory is cloned etc…

dmsnell commented on client/state/domains/management/validation-schemas/reducer.js:
```diff
@@ -0,0 +1,17 @@
+
+export const validationSchemas = ( state = {}, action ) => {
+	switch ( action.type ) {
+		case DOMAIN_MANAGEMENT_VALIDATION_SCHEMA_ADD:
```
can we add some kind of comment or explanation here of what `action.schemas` is expected to look like?

deBhal replied:
✅

deBhal commented:
`dispatchRequestEx`really did clean that up a lot, thanks for that suggestion :)

I've rebased this on master, and rebased #21524 on this, so that PR is probably a better choice for running the code, but you can still verify the redux state with the devtools in this branch if you prefer.

With the multiple requests, it only makes any difference when a user has more than one tld in their cart, which is something we've discussed as being rare enough that we don't want to spend a lot of time supporting it, we just want it not to explode, so I was all set to not just leave the single endpoint, but I tripped myself with the singular/plural separation a couple of times in the rebase, so I think now that I will update the endpoint to accept multiple tlds now before anything's using it.

TODO:
- [ ] Update the API endpoint to accept multiple tlds
- [ ] Update this PR to match

deBhal commented:
@dmsnell: Ok, back end updated to accept multiple comma separated PRs, and this PR updated to match

dmsnell commented on client/state/data-layer/wpcom/domains/validation-schemas/index.js:
```diff
@@ -0,0 +1,64 @@
+export const onError = () =>
+	errorNotice(
+		translate( "We couldn't load the data for validating ccTLD specific requirements." )
+	);
```
This is going to present itself to our customers. I'd like to challenge us to think about whether they would anticipate it and what they might do in response to it. I don't think we should expect someone to know what this is or deal with it.

Instead, could we alert ourselves on the backend and provide a failover behavior for the client? Maybe the network is down, maybe the schema slug is invalid. I'm not sure but what can Calypso do at that point? Maybe we can inject a fake validator? Should we prevent submitting the form?

What if we bumped a stat or fired off a Tracks event recording the requested schema and the failure (maybe the HTTP status code) then injected an `any` type schema for the requested ccTLDs?

```js
export const onError = ( { tlds } ) => [
	...tlds.map( tld => bumpStat( 'ccTLD_validators', `missing_${ tld }` ),
	addValidationSchemas( fromPairs( tlds.map( tld => [ tld, '{ "type": "object" }' ] ) ) )
];
```

or something like that

dmsnell commented on client/state/domains/management/validation-schemas/reducer.js:
```diff
@@ -0,0 +1,18 @@
+	switch ( action.type ) {
+		case DOMAIN_MANAGEMENT_VALIDATION_SCHEMAS_ADD:
+			// action.schemas should contain schema objects keyed by tld, like:
+			// { uk: ukValidationSchema, fr: frValidationSchema, ... }
```
this is a super-helpful comment - thanks for giving an example of possible input data!

dmsnell commented:
> back end updated to accept multiple comma separated PRs

you are so awesome man. this is great
